source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2075 @@
+
+    aws_mutex_lock(&nw_socket->synced_data.lock);
+    if (s_validate_event_loop(nw_socket->event_loop)) {
+        struct aws_task *task = aws_mem_calloc(nw_socket->allocator, 1, sizeof(struct aws_task));
+        struct socket_close_complete_args *args =
+            aws_mem_calloc(nw_socket->allocator, 1, sizeof(struct socket_close_complete_args));
+
+        args->shutdown_complete_fn = nw_socket->on_socket_close_complete_fn;
+        args->user_data = nw_socket->close_user_data;
+        args->allocator = nw_socket->allocator;
+        args->nw_socket = nw_socket;
+        task->arg = args;
+        aws_ref_count_acquire(&nw_socket->ref_count);
+        aws_task_init(task, s_close_complete_callback, args, "SocketShutdownCompleteTask");
+
+        aws_event_loop_schedule_task_now(nw_socket->event_loop, task);
+    } else {
+        // If we are not on the event loop
+        if (nw_socket->on_socket_close_complete_fn) {
+            nw_socket->on_socket_close_complete_fn(nw_socket->close_user_data);
+        }
+    }
+    s_release_event_loop(nw_socket);
+    aws_mutex_unlock(&nw_socket->synced_data.lock);
+    aws_ref_count_release(&nw_socket->ref_count);
+}
+
+int aws_socket_init_apple_nw_socket(
+    struct aws_socket *socket,
+    struct aws_allocator *alloc,
+    const struct aws_socket_options *options) {
+    AWS_ASSERT(options);
+    AWS_ZERO_STRUCT(*socket);
+
+    // Network Interface is not supported with Apple Network Framework yet
+    size_t network_interface_length = 0;
+    if (aws_secure_strlen(options->network_interface_name, AWS_NETWORK_INTERFACE_NAME_MAX, &network_interface_length)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_SOCKET,
+            "id=%p fd=%d: network_interface_name max length must be %d length and NULL terminated",
+            (void *)socket,
+            socket->io_handle.data.fd,
+            AWS_NETWORK_INTERFACE_NAME_MAX);
+        return aws_raise_error(AWS_IO_SOCKET_INVALID_OPTIONS);
+    }
+    if (network_interface_length != 0) {
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_SOCKET,
+            "id=%p fd=%d: network_interface_name is not supported on this platform.",
+            (void *)socket,
+            socket->io_handle.data.fd);
+        return aws_raise_error(AWS_ERROR_PLATFORM_NOT_SUPPORTED);
+    }
+
+    struct nw_socket *nw_socket = aws_mem_calloc(alloc, 1, sizeof(struct nw_socket));
+    nw_socket->allocator = alloc;
+
+    socket->allocator = alloc;
+    socket->options = *options;
+    socket->impl = nw_socket;
+    socket->vtable = &s_vtable;
+
+    aws_mutex_init(&nw_socket->synced_data.lock);
+    aws_mutex_init(&nw_socket->synced_state.lock);
+
+    aws_mutex_lock(&nw_socket->synced_data.lock);
+    nw_socket->synced_data.base_socket = socket;
+    aws_mutex_unlock(&nw_socket->synced_data.lock);
+
+    s_set_socket_state(nw_socket, socket, INIT);
+
+    if (s_setup_socket_params(nw_socket, options)) {
```

Comment:
This should be moved rigth after the nw_socket is created (after the allocator set).  Otherwise a failure here leaves multiple mutexes un-cleaned-up.